### PR TITLE
Narrow by comparisons to boolean literals

### LINKED ISSUE #31105: Strict equals false type guard doesn't work
**Code**

```ts
interface A {
  a: string;
}

interface B {
  b: string;
}

type MyType = A | B;

function isA(x: MyType): x is A {
  return "a" in x;
}

function someFn(x: MyType) {
  if (!isA(x)) {
    console.log(x.b);
    //            ^ <- works perfectly
  }

  if (isA(x) === false) {
    console.log(x.b);
    //            ^ <- here's the problem
  }
}
```

**Expected behavior:**

Being able to access `x.b` in this if statement:

```ts
if (isA(x) === false) {
  console.log(x.b);
}
```

**Actual behavior:**

Error: Property b does not exists on type `A`.

**Playground Link:**

[Playground](https://www.typescriptlang.org/play/index.html#src=interface%20A%20%7B%0D%0A%20%20a%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20B%20%7B%0D%0A%20%20b%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Atype%20MyType%20%3D%20A%20%7C%20B%3B%0D%0A%0D%0Afunction%20isA(x%3A%20MyType)%3A%20x%20is%20A%20%7B%0D%0A%20%20return%20%22a%22%20in%20x%3B%0D%0A%7D%0D%0A%0D%0Afunction%20someFn(x%3A%20MyType)%20%7B%0D%0A%20%20if%20(!isA(x))%20%7B%0D%0A%20%20%20%20console.log(x.b)%3B%0D%0A%20%20%20%20%2F%2F%20%20%20%20%20%20%20%20%20%20%20%20%5E%20%3C-%20works%20perfectly%0D%0A%20%20%7D%0D%0A%0D%0A%20%20if%20(isA(x)%20%3D%3D%3D%20false)%20%7B%0D%0A%20%20%20%20console.log(x.b)%3B%0D%0A%20%20%20%20%2F%2F%20%20%20%20%20%20%20%20%20%20%20%20%5E%20%3C-%20here's%20the%20problem%0D%0A%20%20%7D%0D%0A%7D)

### LINKED ISSUE #44366: typeguards only narrow types when used with unary not
# Bug Report

### 🔎 Search Terms

- negation
- typeguard
- unary not

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about narrowing/typeguards

The FAQ entry with the most promising title is a `TODO`: https://github.com/Microsoft/TypeScript/wiki/FAQ#why-doesnt-isfoox-narrow-x-to-foo-when-isfoo-is-a-type-guard

This occurs on every version of Typescript available in the playground, including nightly.

### ⏯ Playground Link

[Playground Link](https://www.typescriptlang.org/play?ts=4.3.2#code/JYOwLgpgTgZghgYwgAgKLmGAnsg3gKGWWwAcIAuZAZzClAHMBufAX33wQHsQbkBBAMIAVAPIAlAPpCAmgAVUyALzIA5IjCcoK5vlCRYiFHwQaoyCAA9IIACZU0GbHkLEsZSqQicY-YeKlyqMxs+DAAriAmwNzIwFTGpgAUEI5YlOhgmFgAlJQpmU5x-CaazkRQEGBhUCDmqQB0nkqKyoKikjLywexcPGDIMJyclAmlyrielGolWiyMyAD0C8gAbnAANsA2ADTIcPaWZCYQNhzcvABGcFAjM0p4k6ognGAAtGAAFnBgKnOLy6A1psdnsDhYjpBTr1LqB0ql7hM3BRVJgIABbX7zJarDZbXbPfpwWrqTTsYA+RJxUZQRIXUDZZrKeDrKgQBkEIiebzIOkgLEA+wZLK7fbmcEQY6nFjmFkoDmuMjc3n82KC1L4l5iiEnVhkilUma0+nNAYbVnslxcnzK1gy1llBVea2gFVFIXYXZUOBolCiuAXTgrCC63QUgCEBqSvOyFs5SKVLv+qocBSwIrB2qldrllvjzr5SaK1PTWolkNYQA)

### 💻 Code

```ts
interface Entity {
  type: string;
}

const ACTOR_TYPE = 'actor';

interface Actor extends Entity {
  type: typeof ACTOR_TYPE;
}

function isActor(entity: Entity): entity is Actor {
  return entity.type === ACTOR_TYPE;
}

const foo: Actor = {type: 'actor'}; // valid, as expected
const bar: Actor = {type: 'not-that'}; // invalid, as expected
const bin: Entity = {type: 'item'}; // valid, not an actor

if (isActor(bin) === false) {
  typeof bin; // is Entity, as expected
} else {
  typeof bin; // is Entity, not as expected
}

if (isActor(bin) == false) {
  typeof bin;
} else {
  typeof bin; // unexpected Entity, same as above
}

if (!isActor(bin)) {
  typeof bin; // is Entity, as expected
} else {
  typeof bin; // is Actor, as expected
}
```

### 🙁 Actual behavior

Comparing the result of a typeguard with `guard(it) === false` or `== false` does not narrow the type, but the unary operator with `!guard(it)` does narrow it to `Actor`.

### 🙂 Expected behavior

This may be due to a subtlety with the `entity is Actor` return type not being a proper boolean, but I would expect all forms of negation and comparison to `false` to narrow the type, and https://www.typescriptlang.org/docs/handbook/2/narrowing.html#equality-narrowing seems to support that. For a typeguard that returns a boolean, the comparisons should be equivalent. Is there some reason that is not the case here?

I'm working with lint rules that discourage unary not (easy to miss) and discovered this while doing some refactoring.

### PR DESCRIPTION
**Pull request by Andarist**
fixes #31105
closes #44366

### CODE REVIEW

**Comment by jakebailey** [comment]
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this faster

**Comment by jakebailey** [comment]
@typescript-bot pack this

Need to minimize an RWC change.

**Comment by jakebailey** [comment]
This is the reduced diff from the RWC case:

https://www.staging-typescript.org/play?ts=5.1.0-pr-53714-11#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXywGcB5AI0JBgDcpSIQyArcDACgwAstUBzALnhRUATwCUAztx4FC8AJJkK1WvSYsA3AFgAUDu4ZKiKGAQLylGnQalmmeAG8d8Z-ABEOc8qsBaD7YyuAmZKlqo2GjoAvjo6IAAeAA44MBhIaJi4+EQAwjgAtgnIBsCsNBDIIAJCwgA08AkwOAmUGMIA-AKEGDBS4vCkODj0Qg5OLliI8KVQ5QgAvAvwqMgQEPAAPuvwZRXwC3PwaKCI3CDAovBwGMgw+EYQFGPOE1MNTS3C8ACEi0cgJ6gzhdHNoXGCCJNWERFBYVNZ-NNZhd9kgZhQLlcbnc0SAnuCXqwvjsQAA6AAkeQ8cRJxMIAG03s0UsIALoYkDXW6oh640Hg5yYrndCp46J8gUcrHwYW8yJAA

**Comment by Andarist** [comment]
@jakebailey I'd consider this an improvement because we get an error **today** here:
```diff
 export function isComputed(value: any, property?: string): boolean {
     if (value === null || value === undefined) return false
     if (property !== undefined) {
-        if (isObservableObject(value) === false) return false
+        if (!isObservableObject(value)) return false
         if (!value.$mobx.values[property]) return false
         return true
     }
     return true
 }
```

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -26992,6 +26994,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return type;
         }
 
+        function narrowTypeByBooleanComparison(type: Type, expr: Expression, bool: BooleanLiteral, operator: BinaryOperator, assumeTrue: boolean): Type {
+            assumeTrue = (assumeTrue !== (bool.kind === SyntaxKind.TrueKeyword)) !== (operator !== SyntaxKind.ExclamationEqualsEqualsToken && operator !== SyntaxKind.ExclamationEqualsToken);
+            return narrowType(type, expr, assumeTrue);
```
Comment (Andarist) [inline src/compiler/checker.ts:26999]: I probably should take another look at this and conditionally use `narrowTypeByEquality` or something like that. Gonna investigate this 🔜

**Comment by Andarist** [comment]
I came back to this after a while and I think that this PR is in good shape and can be reviewed.

There are potentially some issues here related to coercion but this **isn't** a new problem:
```ts
function test(a: string | number | boolean) {
  if (a == false) {
    a; // actual: false, expected: "" | 0 | false
  }
}
```

So I don't think that those should block this PR.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -27610,6 +27612,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return type;
         }
 
+        function narrowTypeByBooleanComparison(type: Type, expr: Expression, bool: BooleanLiteral, operator: BinaryOperator, assumeTrue: boolean): Type {
+            assumeTrue = (assumeTrue !== (bool.kind === SyntaxKind.TrueKeyword)) !== (operator !== SyntaxKind.ExclamationEqualsEqualsToken && operator !== SyntaxKind.ExclamationEqualsToken);
```
Comment (jakebailey) [inline src/compiler/checker.ts:27616]: I think my only comment on this bit at this point is just how unreadable this expression is; I honestly have no idea what I'm looking at here. Is there a more obvious way to represent this?

Comment (Andarist) [reply src/compiler/checker.ts:27616]: Iirc this is a double xor with booleans

Comment (jakebailey) [reply src/compiler/checker.ts:27616]: That makes me feel slightly better, but this is still like 3 nested equalities...

Comment (Andarist) [reply src/compiler/checker.ts:27616]: I don't disagree 😅 at first I started with some nested if statements - and believe me... it didn't look good either. Then I realized that what I writing there was really just a xor so I managed to fold it all into this one line. It didn't look that great to me either. I think it's a very nice way of doing this but the readability suffers here if one doesnt know what this is. I tried to give it some names but I don't have any good ones so I gave up. The best alternative that I can offer would be smth like:
```ts
const a = assumeTrue;
const b = bool.kind === SyntaxKind.TrueKeyword;
const c = operator !== SyntaxKind.ExclamationEqualsEqualsToken && operator !== SyntaxKind.ExclamationEqualsToken

// boolean xors
assumeTrue = (a !== b) !== c
```

**Comment by jakebailey** [comment]
@typescript-bot test top200
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this faster

**Comment by jakebailey** [comment]
@typescript-bot pack this

**Review by jakebailey — APPROVED** [review]
Inspecting [Playground Link](https://www.staging-typescript.org/play?noEmit=true&ts=5.3.0-pr-53714-29#code/PTAEAEGcBcCcEsDG0Bco4FcCmAoEEA7AewFEBbeVdWbHHaATwActQBBUAXlAG91msaAERshoAL4BuegNAAhLr34thcsVJktQAYUV9GK0EO3rpB1gFkGAVQLwiBRRwA+80K+3SciBzFDxIDm4ACgAPNCtbewIASjRQ-0h2LgA+UFCAOnMuTm4RIS8AMwwCZGj0LBgARjCImzsHGN4cUFb-QtBggLYwpoBCXOpsJp4WtvHQ6XH8UAA9AH4x1vE6cfgOrsDe0AHuQoBDABtILBGl8fSptpmF85Xz9c7u7dy9o5Ozi7bJ85vF8fuaw2z1CTVeQ1OzS+rR+0zAc3+bUBbUemx6oJ23EwkNG0Muv3htwBqxRwK2GMG2M+0Nh10JiOWJNaqOxmO4IJi1K+tNafzuTPanVZ4I5XIuPNAfOJ93wAAtoNAmJAUCAAOaUWUYABGGR8ZGAFEQsCIkCIhWgwAAKgIAMpG+BMC0BSDYSDAACsAGYAAwATk9OGKpWg5WglWgACZaqASgBrYgAd1iUNJnQS8AIMH2pSwZtAJFgxtgOTexxx53GsCw0AwsAIV0ZE2kfJlYHliuVao12t1RH1huNpvNVtt9sdwGdro93oA7FUI4GSmUHBUYJ7goUiEQ0HHE8ncanguY85uiGzQAByGAIAiqi+gABkD-Yhf2DAy3VfDA3W7BgwOZZiuM0CysaCagAQWDgQWRbBBeAByRDQKA+ygNeGaqqARDFtmKFfheMQNhI5yns29I4K2oDtkqKrAOqIE9nqBpIIOZoWtaLB2gg46TpUHper67reIc+yQEkADqWBajB2GgFgoRhgQAAmSTWAASgAkjJxYHq0WY1pA8xoAQGBkFqWCwNIKxBsujhhjAAAswRinADApl8+BkPssasKhVYAI6utAdygIg+zQIgsqdBZsBAYeMX+Jm0DZoguYdJJ0mFrJ4IASc7iuDFGT6RgSQDKADnelUcVfD4mZEIcWAZDF2HBDFhEVtCMx4m0RLQis0p0HKCo0V2DE6kxA4mmxI6cWOTpiVODkOZ6ABsK04BmYawAcqX5gQIaMO5yiCGhcAYVZ3i+MhbDaJaADyakAPqWgAmgACiQihCPsyDYQUG37RZO2sGwv3FvJikqXtB1ubpx1oMeHQ3fdT2vR9F02SGK7dGDrX7ZQDBoCQ+OMHEckk25ATsGDR1VjWdbkzDWSyOCyMPc970kBjS5Y3Z4busEWoZkTFNiqiOPQC1QvJjl7zlni0tEVKSJyWWR3jIrBKtL1jLWTzobhitgvC9DBNi2SoOS7Axsy6WHzq20mtwtrDISKreVwxrGZK+RA14G2w2dnR3bjX2zFGlNw4cVgXEOvNLp8V63rekJmPlAEABi+sONGu5EEmZPpkkWfBuUcN07WdkCHmCTgkIacOP99wN3zMAzrnBDxvn+4PGSJe2dsuygLl8s0j7Lv8kCTyQP3vMvJSNCj9y48Ivy4hAA) (as it's a little easier to read than the baselines), it seems like things are correct.

@RyanCavanaugh @DanielRosenwasser is this something we want for sure?

**Comment by Andarist** [comment]
I think that Ryan confirmed in March that this is smth that the team would accept PR for (see [here](https://github.com/microsoft/TypeScript/issues/53170#issuecomment-1462516268))

**Comment by jakebailey** [comment]
Ah, yeah, you're right.

**Pull request merged**